hector_ros_controllers/src/velocity_to_position_controllers_base.cpp

```diff
@@ -165,68 +263,123 @@ VelocityToPositionControllersBase::on_deactivate( const rclcpp_lifecycle::State
   }
+}
 
-  if ( e_stop_active_ || invalid_state ) {
-    for ( auto index = 0ul; index < command_interfaces_.size(); index++ ) {
-      const auto &state = state_interfaces_[index].get_optional();
-      if ( state.has_value() && !std::isnan( state.value() ) ) {
-        last_positions_[index] = state.value();
-        successful = command_interfaces_[index].set_value( last_positions_[index] );
+void VelocityToPositionControllersBase::update_sync_offsets()
+{
+  for ( size_t joint_idx = 0; joint_idx < joints_.size(); joint_idx++ ) {
+    if ( !sync_states_[joint_idx] ) {
+      for ( size_t i = 0; i < synced_joints_[joint_idx].size(); i++ ) {
+        sync_offsets_[joint_idx][i] =
+            joint_position_states_[synced_joints_[joint_idx][i]] - joint_position_states_[joint_idx];
       }
     }
+  }
+}
+
+double VelocityToPositionControllersBase::sync_p_control( const size_t &joint_idx )
+{
+  double sync_pos_command = 0.0;
+  for ( size_t i = 0; i < synced_joints_[joint_idx].size(); i++ ) {
+    sync_pos_command += ( joint_position_states_[synced_joints_[joint_idx][i]] -
+                          joint_position_states_[joint_idx] - sync_offsets_[joint_idx][i] ) *
+                        kp_sync_;
+  }
+  return sync_pos_command / (double)synced_joints_[joint_idx].size();
+}
+
+double VelocityToPositionControllersBase::pos_pd_control( const size_t &joint_idx,
+                                                          const double &vel_command,
+                                                          const rclcpp::Duration &p )
+{
+  // FF + P + D
+  return joint_position_states_[joint_idx] + vel_command * p.seconds() +
+         kp_ * ( vel_command - joint_velocity_states_[joint_idx] ) * p.seconds() -
+         kd_ * ( joint_velocity_states_[joint_idx] - joint_prev_vel_states_[joint_idx] ) *
+             ( p.seconds() * p.seconds() );
+}
+
+double VelocityToPositionControllersBase::position_control( const size_t &joint_idx,
+                                                            const double &vel_command,
+                                                            const rclcpp::Duration &p )
+{
+  double next_position = pos_pd_control( joint_idx, vel_command, p );
+  if ( sync_states_[joint_idx] )
+    next_position += sync_p_control( joint_idx );
+
+  return next_position;
+}
+
+controller_interface::return_type
+VelocityToPositionControllersBase::update_and_write_commands( const rclcpp::Time & /*time*/,
+                                                              const rclcpp::Duration &p /*period*/ )
+{
+  update_joint_states_if_valid();
+  if ( !interfaces_valid_ )
+    return controller_interface::return_type::ERROR;
+
+  bool successful = true;
+  if ( e_stop_active_ ) {
+    for ( auto index = 0ul; index < command_interfaces_.size(); index++ ) {
+      if ( !std::isnan( joint_position_states_[index] ) )
+        successful = command_interfaces_[index].set_value( joint_position_states_[index] );
```

Comment:
Currently, only the last iteration in the loop affects success.
```suggestion
        successful &= command_interfaces_[index].set_value( joint_position_states_[index] );
```

Reply:
Decided to remove any command writing if e_stop is active